kernel/src/deferred_call.rs

```diff
@@ -1,84 +1,231 @@
-        AtomicUsize {
-            v: UnsafeCell::new(v),
-        }
-    }
+// This trait is not intended to be used as a trait object;
+// e.g. you should not create a `&dyn DeferredCallClient`.
+// The `Sized` supertrait prevents this.
+/// This trait should be implemented by clients which need to
+/// receive DeferredCalls
+pub trait DeferredCallClient: Sized {
+    fn handle_deferred_call(&self);
+    fn register(&'static self); // This function should be implemented as
+                                // `self.deferred_call.register(&self);`
+}
 
-    pub(crate) fn load_relaxed(&self) -> usize {
-        unsafe { intrinsics::atomic_load_relaxed(self.v.get()) }
-    }
+/// This struct serves as a lightweight alternative to the use of trait objects
+/// (e.g. `&dyn DeferredCall`). Using a trait object, will include a 20 byte vtable
+/// per instance, but this alternative stores only the data and function pointers,
+/// 8 bytes per instance.
+#[derive(Copy, Clone)]
+struct DynDefCallRef<'a> {
+    data: *const (),
+    callback: fn(*const ()),
+    _lifetime: PhantomData<&'a ()>,
+}
 
-    pub(crate) fn store_relaxed(&self, val: usize) {
-        unsafe { intrinsics::atomic_store_relaxed(self.v.get(), val) }
+impl<'a> DynDefCallRef<'a> {
+    // SAFETY: We define the callback function as being a closure which casts
+    // the passed pointer to be the appropriate type (a pointer to `T`)
+    // and then calls `T::handle_deferred_call()`. In practice, the closure
+    // is optimized away by LLVM when the ABI of the closure and the underlying function
+    // are identical, making this zero-cost, but saving us from having to trust
+    // that `fn(*const ())` and `fn handle_deferred_call(&self)` will always have the same calling
+    // convention for any type.
+    fn new<T: DeferredCallClient>(x: &'a T) -> Self {
+        Self {
+            data: x as *const _ as *const (),
+            callback: |p| unsafe { T::handle_deferred_call(&*p.cast()) },
+            _lifetime: PhantomData,
+        }
     }
+}
 
-    pub(crate) fn fetch_or_relaxed(&self, val: usize) {
-        unsafe { intrinsics::atomic_store_relaxed(self.v.get(), self.load_relaxed() | val) }
+impl DynDefCallRef<'_> {
+    // more efficient pass by `self` if we don't have to implement `DeferredCallClient` directly
+    fn handle_deferred_call(self) {
+        (self.callback)(self.data)
     }
 }
 
-unsafe impl Sync for AtomicUsize {}
+// The below constant lets us get around Rust not allowing short array initialization
+// for non-default types
+const EMPTY: OptionalCell<DynDefCallRef<'static>> = OptionalCell::empty();
 
-static DEFERRED_CALL: AtomicUsize = AtomicUsize::new(0);
+// All 3 of the below global statics are accessed only in this file, and all accesses
+// are via immutable references. Tock is single threaded, so each will only ever be
+// accessed via an immutable reference from the single kernel thread.
+static mut CTR: Cell<usize> = Cell::new(0);
```

Comment:
Can you add a doc comment explaining what the `CTR` counts? Similarly, a quick explanation of the bits of `BITMASK` and how they relate to the entries in `DEFCALLS`

Reply:
Done